### ISSUE TITLE
TF VPG fails (on Cartpole)

### ISSUE THREAD

**Issue by andreasdoerr** [rlworkgroup/garage#417]
I was trying to run the Tensorflow implementation of VPG on Cartpole as a sanity check and to reproduce the results reported in the "Benchmarking Deep Reinforcement Learning for Continuous Control" paper.

Learning fails for both, the configuration reported in the paper (using multiple random seeds) and for a simpler policy network (hidden size: 32, 32). Average return drops from ~40 for the initial policy to ~20 and never recovers. The Theano implementation seems to work just fine.

Minor comment: 
In https://github.com/rlworkgroup/garage/blob/master/garage/tf/algos/vpg.py it should be
`pg_loss=PGLoss.VANILLA,` 

My simple start script to run VPG on Cartpole and reproduce my problem is:
Am I missing something? (For theano I only replaced VGP/TfEnv and GaussianMLPPolicy with their Theano counterparts).

```python
from garage.baselines import LinearFeatureBaseline
from garage.misc.instrument import run_experiment
from garage.tf.algos import VPG
from garage.tf.envs import TfEnv
from garage.envs import normalize
from garage.envs.box2d import CartpoleEnv
from garage.tf.policies import GaussianMLPPolicy

def run_task(*_):

    env = TfEnv(normalize(CartpoleEnv()))
    
    policy = GaussianMLPPolicy(
        name="policy", env_spec=env.spec, hidden_sizes=(100, 50, 25))
    
    baseline = LinearFeatureBaseline(env_spec=env.spec)
    
    algo = VPG(
        env=env,
        policy=policy,
        baseline=baseline,
        batch_size=50000,
        max_path_length=500,
        n_itr=500,
        discount=0.99,
        optimizer_args=dict(tf_optimizer_args=dict(learning_rate=5e-3, )))
    
    algo.train()
    
    
run_experiment(
        run_task,
        n_parallel=1,
        snapshot_mode="last",
        seed=1)
```

**Comment by andreasdoerr** [comment]
Thanks for this fix. Both Theano and Tensorflow versions of VGP now seem to work. However, I still get (very) different behaviour/results.

Using the script from above (VGP on the Cartpole environment):
* Theano-VGP converges slower compared to TF-VGP (20 steps vs 70 steps to reach optimal performance)
* Final performance of Theano-VGP is better than TF-VGP (4800 avg. return vs. 4500 avg. return)

Second thing I noticed:
In both implementations, I consistently get rollouts with 0 min. return even for well trained policies (for the cartpole environment this means rollouts of length 0/1). Could it be that the sampler is somehow messing up?

**Comment by CatherineSue** [comment]
@andreasdoerr Np. Thanks for pointing out the bug.
First thing, I tested Theano/VPG and TF/VPG using the same hyperparameters above, the final performance of TF/VPG is better than Theano/VPG (5000 avg. return vs 4901 avg. return). The Theano tree uses a different sampler than TF, which may relate to the converge steps. We recommend users to use TF tree as it is known to have a better performance. 

Second thing, a path’s total reward is 0 happens when the rollout path length is 1. The possible reason for the minReturn is constantly 0 is that the batch_size is too high, which is 50000. It is highly possible that the current observation and action will go to a terminal state as the exploration space is big. If we decrease the reset_range in garage.envs.box2d.CartpoleEnv.reset(), the rollout path in average is gonna be longer.

**Comment by ryanjulian** [comment]
@CatherineSue is this fixed?

**Comment by CatherineSue** [comment]
Yes
